GLOBAL_DEFINED_WORDS.json

```diff
@@ -16,6 +16,10 @@
   },
   "未定義動作": { "redirect": "未定義の動作", "yomi": "みていぎどうさ" },
   "未定義": { "redirect": "未定義の動作", "yomi": "みていぎ" },
+  "未定義アドレス": {
+    "yomi": "みていぎあどれす",
+    "desc": "配列要素へのポインタがアドレス計算の結果として配列の範囲外に到達した場合に、(間接参照しなくても) 未定義動作となる"
```

Comment:
「未定義アドレス」についてはこんな感じの対応を想定してました。
```suggestion
    "desc": "通常は未定義動作となるポインタの加減算が atomic 操作として行われた場合の結果を指す規格中の表記 "undefined address" を指す。その意味付けは不明確なので LWG issue 3906. "Undefined address" is undefined が提起されている。",
    "link": "https://cplusplus.github.io/LWG/issue3906"
```

Reply:
> atomic ポインタ操作について「未定義アドレス」が現れる規定の意図は、通常のポインタ演算で生じる 上記のような未定義動作が起こらないとすることだったので、これは間違った情報となってしまいそうです。 https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2393.html
> 
> > The normal signed integral addition operations have undefined behavior in the event of an overflow.
> > For atomic variables, this undefined behavior introduces significant burden because there is in general
> > no way to pre-test for overflow. ...

全部読んだわけではないですが、上記の部分は符号付き整数についての記述で、一方で該当の未定義アドレスの部分の文章は以下のようになっています。

> For signed integral types, arithmetic is defined to use two's-complement representation. There are no undefined results. For address types, the result may be an undefined address, but the operations otherwise have no undefined behavior.

符号付き整数とアドレスの対比で書かれているこの記述を見ると、少なくともこの文書の著者は "未定義アドレス" の語を未定義動作を意図して使っているように読めます。(もし未定義アドレスがアドレスでも "未定義動作が起こらない" ようにするために導入された語であるとするならば、上記のようにわざわざ書く必要はなく、アドレスについても単に There are no undefined results. と書けば良いので、ここでわざわざ未定義アドレスという語を導入するのは不自然である。)

> 「未定義アドレス」の後につづく記述「それ以外の未定義動作はない」の「それ以外の」も同様に
> 結果が「未定義アドレス」となる場合に未定義動作となるという解釈を示すものとなりそうなので、
> 削って単に「未定義動作はない」としておいたほうがよさそうです。

なので、この特定箇所の説明でのこの解釈自体は妥当なものだと思います。むしろ「未定義動作はない」というのは違うのではないかと思います。ただ、読者がこの記事を見て「未定義アドレスは一般に未定義動作を表すものだ」という印象を持つ可能性があるのは問題かもしれません。記事内で「未定義アドレスの意味について LWG issue 3906 が提起されている」ことも記述しても良いのではないかと思います(単なるポップアップではなく)。

Reply:
うーん。用語が不明瞭なのでそういう解釈もできちゃいますかね。
僕のほうの解釈に至る理由をいくつか追加で挙げてみます。

[N2393](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2393.html) 冒頭近くには整数かポインタかを限定しない以下の記述もあり、ここからも
通常の演算で起こる未定義動作を atomic 操作全体では起こらないように定める意図ががあったと見てます。
> Specifically, a program that ..., needs a shared variable flag, that both:
> - Allows atomic accesses, in the sense that concurrent reads and writes are allowed and that the reads result in only one of the assigned values and never undefined behavior.

また、符号付き整数について未定義動作が起こらないとされた理由
"... because there is in general no way to pre-test for overflow" はポインタにも
同様に当てはまるものと思います。

> (もし未定義アドレスがアドレスでも "未定義動作が起こらない" ようにするために導入された語であると
> するならば、...、アドレスについても単に There are no undefined results. と書けば良いので、
> ここでわざわざ未定義アドレスという語を導入するのは不自然である。)

通常の演算では未定義動作とするケースで未定義動作が起こらないとする場合、該当のケースでは
どうなるのかを定める必要が出てきます。なので単に There are no undefined results. と書くだけでは
済まず、何かしら結果を定めるための文言が入っているのは不自然とは思ってません。
そこでアドレス値が範囲外や別のオブジェクトを指すことになるような（単純な実装で生じる）結果を指して
"undefined address" がカジュアルに生まれて、当時の論文読者にぼんやりと意味づけは伝わって
問題とされなかったんだろうというのも、ありそうな話として理解できます。

むしろ通常の演算と同様に未定義動作となるのであれば Remarks 全体が不要になるので、
そのほうが不自然とも思います。

Reply:
あとは改めて探して見つけたものですが、 C でもこの部分は同じ文言があって不明瞭さが
問題とされており、そこでも未定義動作は起こらないことを明確にする方向で論文が出ていました。
https://www.open-std.org/JTC1/SC22/WG14/www/docs/n1955.htm (2015)
> Conclusion
> ... They would better be stated as such.
>   ...
>   2. Pointer arithmetic must have a variant that always has defined behavior, only that the stored address may be invalid, if the addition or subtraction passed beyond the boundaries of the object. ...
>   ...

この件について最新は具体的な修正案を含む (WG14)[N2389](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2389.pdf) (2019) となり、まだ採択はされてないようですが、
上記の方針は変わっておらず、結果は "unspecified" にしようとしている模様。
（↓ "If C2 is a pointer ..." ２つ目の "indeterminate" は "undefined" の間違いかな？）
> 6.5.16.2 Compound assignment
> ...
> - If C2 is a pointer type and the value of old is null or is indeterminate, the value of new is unspecified.
> - If C2 is a pointer type and the value of " old op val" would be indeterminate, the value of new is unspecified.

Reply:
お返事遅れましてすみません。確かに言われるとそうだと思いまして、一方で腑に落ちないこともいくつかあって時々考えていたのですが、たぶん元の文書では意識的な無意識的かうやむやにして誤魔化したというか誤魔化された感じだったのかなという気がしています。

腑に落ちなかったことと言うのは、

- まず、元の N2397 では終始符号付き整数のUBの問題を議論していて範囲外ポインタのUBのことは考慮していない様に思えるということ。例えば、動機のところの "this undefined behavior introduces significant burden" の文脈では符号付き整数の overflow によって起こる UB の話であって一般の UB の話をしている訳ではありません。 
- また、符号付き整数型については、UBを取り除くことができる理由としてわざわざ処理系として内部表現が2の補数で変なことが起こらないものしか現実的に考えなくて良いという事が述べられている一方で、ポインタに関しては何も述べられていないこと。符号付き整数についてはUBを除けるか慎重に理由を与えているのに、ポインタについては何も言及せずに俎上にないUBを除いたと考えるのはいかにも不自然です。
- それにやはり "For address types, the result may be an undefined address, but the operations otherwise have no undefined behavior." の "otherwise" はそれより前で述べていることがUBということを含意しています。未定義のアドレスという語を以てUBが起こらないという意図だとすると、英文として変です。

だけど最後の点については「アドレス型については、未定義のアドレスになる場合があって (その後に指し示す先を読み取ろうとした時にUBになるが)、それ以外の場合にはUBはない」と補って解釈すれば、いいかげんな英文としてまあなくはないかもという気がしてきました。

総じて受ける印象は N3279 の時点では範囲外ポインタのUBのことを意識していなかった(考えるのを忘れていた or 未定義アドレスは間接参照するまでUBではないと思っていた?) のではないかということです。そんな感じでうやむやな感じで通った文書ではないかと。でも、改めて書いていてやはり思うのですが「N3279 の時点で範囲外ポインタのUBを意識的に除外した」と解釈するのには無理があるように思います。後の文書によってその辺りについて顧みられていたとしても。

Reply:
@akinomyoga ↑を受けての結論がよくわからなかったので確認させてください。
上記の理由により、やっぱり腑に落ちないから単に「未定義動作はない」としてしまうことには反対ということで合ってますか？

それとも、 [N2393](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2393.html) 当時からの意図だったというのは腑に落ちないけど、その他の C(WG14) の動きなども併せれば
「未定義動作はない」としておくのがよさそうと合意できた（＝「そうだと思いまして」）ということでしょうか？

Reply:
Cの方に提案された変更では、たしかに以下のようにありますね。

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2393.html
> Pointer arithmetic must have a variant that always has defined behavior, only that the stored address may be invalid, if the addition or subtraction passed beyond the boundaries of the object. But that behavior is not defined by the standard, the negation of undefined doesn't give a definition.
> 
> ポインタ算術は常に定義された動作のバリアントをもたなければならない。ただし、加算や減算がオブジェクトの境界を越えた場合、格納されたアドレスは無効となる可能性がある。
> しかし、その (境界を超えた場合の) 動作は標準では定義されておらず、undefinedの否定は定義を与えていない。

アトミックオブジェクトは未定義動作を含まないよう意図されて設計されており、undefined addressとは境界を超えた場合に未規定だが無効のアドレスになる、くらいの意図だったように読み取れます。